### PR TITLE
[Merged by Bors] - Implement iter() for mutable Queries

### LINKED ISSUE #753: Allow immutable iter on Queries which contain mutable references
**What problem does this solve or what need does it fill?**

The new Query api follows rust conventions by using .iter() to create iterators over queries of immutable component references and .iter_mut() to create iterators over queries of mutable component references. However, it deviates from rust convention by not allowing iter() to be called on queries of mutable component references.
 
**Describe the solution would you like?**

I expect .iter() on a mutable iterable object to return an iterator of immutable references rather than fail with a compilation error. That would make the API consistent with other similar APIs in rust and allow programmers to signal clearly where they are iterating over a query which is used mutably elsewhere in the system but is being used immutable at the moment.

**Describe the alternative(s) you've considered?**

I have yet to encounter a case where calling iter_mut is a serious problem, so keeping the existing API is certainly an alternative.

### PR DESCRIPTION
**Pull request by Guvante**
A sample implementation of how to have `iter()` work on mutable queries without breaking aliasing rules.

# Objective

- Fixes #753

## Solution

- Added a ReadOnlyFetch to WorldQuery that is the `&T` version of `&mut T` that is used to specify the return type for read only operations like `iter()`.
- ~~As the comment suggests specifying the bound doesn't work due to restrictions on defining recursive implementations (like `Or`). However bounds on the functions are fine~~ Never mind I misread how `Or` was constructed, bounds now exist.
- Note that the only mutable one has a new `Fetch` for readonly as the `State` has to be the same for any of this to work

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
bors try

**Comment by Guvante** [comment]
Note that I believe #1301 is calling for `iter()` to work in place of `iter_mut()` which is different than the way this PR handles things.

**Comment by Guvante** [comment]
Thought this was going to be a normal PR since GitHub was being odd with the button...

Left as draft due to:

* Not liking `QF` (QueryFetch) as a name where `F` (Filter) is in scope and a different name.
* Unsure how to handle the private member functions that are now generic. In the interest of minimizing changes I swapped them to `Q::Fetch` or `Q::ReadOnlyFetch` depending on whether `ReadOnlyFetch` was a constraint on `Q::Fetch`.
* I left `ReadOnlyFetch` which seems unnecessary.
* Most importantly I didn't touch anything other than what was necessary to relax the `Q::Fetch: ReadOnlyFetch` constraint on `iter()` and I don't think that is the only thing that could use a change like this if it is useful.

**Comment by mockersf** [comment]
I'm not sure this is something we want, as it would encourage using a mutable query when an immutable would work, and mutable have a heavier cost for system parallelisation. Having to use `iter_mut` works a second level of "are you sure?" when you don't want to mutate something

File: crates/bevy_ecs/src/system/query.rs
Comment (NathanSWard) [inline crates/bevy_ecs/src/system/query.rs:564]: If this PR lands, I'm curious if we actually want to replace the implementation of `is_empty` with this. 
Now that I think of it, having a specialized `is_empty` functionality actually gives us a small performance win (even if it is at the cost of extra code duplication). Since in the implementation we never have to touch `self.fetch` so we bypass that extra unnecessary code path. 

Where as in here, we'll have to follow the normal code path and do a bit of extra work.

Comment (Guvante) [reply crates/bevy_ecs/src/system/query.rs]: I refactored this to use a `NopFetch` to get the performance benefit without the code duplication. How does that look?

File: crates/bevy_ecs/src/query/fetch.rs
Comment (Guvante) [inline crates/bevy_ecs/src/query/fetch.rs:1142]: Actually re-reading the usages I think this should be `true`

**Comment by Guvante** [comment]
We and applied the refactor to all of the existing locations. Now the only reference to `ReadOnlyFetch` the trait is in `WorldQuery` and the many `impl`s.

I tried to transform the non-`pub` "_manual" methods to take generics so the API isn't impacted beyond the names of types that now need to know which of the two `WorldQuery::Fetch` you are using.

File: crates/bevy_ecs/src/query/iter.rs
Comment (Guvante) [inline crates/bevy_ecs/src/query/iter.rs:363]: I believe that `QF: Fetch<'s>` and dropping the `Q` and `'w` would work for this, as I believe the lifetime parameter on `Fetch` is "lasts at least as long" but I wasn't 100% sure so went with a more obviously "correct" solution of including them with a PhantomData.

**Comment by alice-i-cecile** [comment]
bors try

**Review by alice-i-cecile — APPROVED** [review]

**Comment by Guvante** [comment]
Let me know if it needs to be repeated on the other issue but:

I license past and future contributions under the dual MIT/Apache-2.0 license, allowing licensees to choose either at their option.

**Comment by alice-i-cecile** [comment]
> Let me know if it needs to be repeated on the other issue but:

> I license past and future contributions under the dual MIT/Apache-2.0 license, allowing licensees to choose either at their option.

@DJMcNab IIRC we wanted this all in one thread? This should suffice, but I'll leave it up to your judgement there :)

**Comment by DJMcNab** [comment]
@cart was relatively clear that we should have them in #2373 as a policy

That being said, if we can't reach this user again then this would be fine.

**Review by alice-i-cecile — APPROVED** [review]
This looks fantastic; that's a nice batch of changes! I've done a thorough review of the code, but I could not spot any particular nits to pick.

It may be nice to have a test to verify that this new functionality continues to work, but I'll leave that to your judgement.

**Comment by Guvante** [comment]
Commented on linked issue as well. I can throw together a unit test that verified `iter()` is a view on the object returned by `iter_mut()`. If you want to double check that no `pub` things expose `QF` that would be great as while I could see exposing that functionality I did not want to do that here.

File: crates/bevy_ecs/src/query/iter.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/iter.rs:13]: Public QF.

File: crates/bevy_ecs/src/query/iter.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/iter.rs:158]: Public QF.

**Comment by alice-i-cecile** [comment]
>  If you want to double check that no pub things expose QF that would be great as while I could see exposing that functionality I did not want to do that here.

Looks like it's exposed in `QueryIter` and `QueryCombinationIter`. I don't immediately see a way to avoid that; why would you want to keep this fully private?

**Comment by Guvante** [comment]
I am worried about a user constructing with a custom `QF`, I don't care that users can know something is generic over a `QF`. Both the `QueryIter` and `QueryCombinationIter` have a `pub(crate)` constructor (new function? not sure the terminology there).

I haven't done the mental to work to decide what it means to have a user supply a custom `QF` to something. Mostly because I would need to figure out what if anything now needs to be market `unsafe`. In theory the need for a matching `QueryState` as well as `ReadOnlyFetch` eliminates most problems but I didn't want to expand the public API options without verifying that is enough.

**Comment by Guvante** [comment]
Rebased with a minor tweak to preserve an existing helper method (previously removed State::is_empty as redundant, now it is used)

**Comment by alice-i-cecile** [comment]
Fixes #2744.

**Comment by Guvante** [comment]
I am unsure what the build failure is telling me, is it a CI bug?

**Comment by cart** [comment]
Its a transitory crates.io error. Nothing to fix, we'll just need to retry before merging.

**Comment by alice-i-cecile** [comment]
bors try

(This will retrigger the CI so we can check again.)

**Comment by alice-i-cecile** [comment]
bors try

**Review by alice-i-cecile — APPROVED** [review]
I've given this another look over and it looks ready to merge. Very nice usability / correctness win, and the docs and tests are solid.

One minor nit, but feel free to merge without it being addressed.

File: crates/bevy_ecs/src/system/query.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/query.rs:965]: We should replace this with the code in the comment now that we've figured out that conversion. I expect it will be both cleaner and marginally faster.

Comment (Guvante) [reply crates/bevy_ecs/src/system/query.rs:965]: `QueryState::is_empty` existed and got updated so I left this as a pass through to it to avoid changing any `pub fn`

Also the implementation isn't quite what is described in the comment as the actual implementation uses a custom `Fetch` to avoid touching any of the data much like the custom `is_empty` did (`NopFetch`)

Comment (alice-i-cecile) [reply crates/bevy_ecs/src/system/query.rs:964]: Sounds good to me.

**Comment by deontologician** [comment]
Any ETA on merging this?

**Comment by deontologician** [comment]
So, I don't have an informed opinion on whether this PR is the right fix for bevy architecturally etc, but I did manage to compile this PR of fixes for bevy_rapier by @pcwalton:
https://github.com/dimforge/bevy_rapier/pull/98
 and it seems to work.

I had to add some things to deal with the new `Component` requirement since that popped up since the original `bevy_rapier` fix branch was made, but I can say that it seems like this PR fixes the issues `bevy_rapier` had compiling against `bevy/main`

File: crates/bevy_ecs/src/query/fetch.rs
Comment (Guvante) [inline crates/bevy_ecs/src/query/fetch.rs:51]: This is a terrible hack...

File: pipelined/bevy_render2/src/render_component.rs
Comment (Guvante) [inline pipelined/bevy_render2/src/render_component.rs:44]: It is late and I can't figure out how to satisfy this bound.

Basically we can't ensure `QueryItem` is returned by `iter` due to needing to allow `Item` to be different to properly support `WriteQuery`... Since it returns `&T` or `Mut<T>` depending on whether you call `iter` or `iter_mut`.

I briefly considered making it work with a singular return type but I don't believe there is a way to do that without a footgun. There isn't a singular type that encapsulates those two types without runtime checks which is a terrible setup.

I believe if `QueryItem` is the way forward the best solve is to keep the idea of two types (`QueryItem` and `ReadOnlyQueryItem`) but name them in a not terrible way.

**Review by Guvante — COMMENTED** [review]
Noticed a conflict which was easy to resolve (if monotonous) however upon building the new `ExtractComponent` is super unhappy with how I broke `QueryItem<T>`...

I really wish there was a way to guarentee that when `Query: ReadOnlyQuery` then the two types are the same but I am almost certain that is an impossible ask.

I can totally ensure that their `Item` values are the same (as I did with `State` just weirder bounds) but that doesn't work due to `&T` vs `Mut<T>` for `WriteQuery`...

**Review by cart — COMMENTED** [review]
This generally looks good to me. Impressive first PR! The extra QF type complexity in our internal apis is a bit painful, but I can't think of alternatives that don't involve code duplication. And QF does give us additional flexibility, which has already yielded fruit (ex: the NopFetch impl). We might even find more create uses of this pattern in the future. I've resolved a couple of my comments already, which I'll push after leaving this review.

File: pipelined/bevy_render2/src/render_component.rs
Comment (cart) [reply pipelined/bevy_render2/src/render_component.rs]: I think the best solution (at least for now) is to just use `iter_mut()` in the ExtractComponent impl and remove the ReadOnlyFetch constraint. It kind of punts this problem forward, but it is niche enough that we might never need to care. And that way we can remove the extra ReadOnlyQueryItem type, which does feel a bit odd. I pushed a commit that does this.

Comment (cart) [reply pipelined/bevy_render2/src/render_component.rs]: The ReadOnlyFetch constraint wasn't needed in this case anyway. I think that was added to accommodate an older design.

File: crates/bevy_ecs/src/query/fetch.rs
Comment (cart) [reply crates/bevy_ecs/src/query/fetch.rs]: Haha I removed it. See my comment below for rationale.

File: crates/bevy_ecs/src/query/filter.rs
Comment (cart) [reply crates/bevy_ecs/src/query/filter.rs]: I changed this to `$filter::ReadOnlyFetch` and added a `$filter::ReadOnlyFetch: FilterFetch` constraint to make that possible.

File: crates/bevy_ecs/src/query/filter.rs
Comment (cart) [reply crates/bevy_ecs/src/query/filter.rs]: I added a ReadOnlyFetch constraint to `$filter` here which ensures we only impl ReadOnlyFetch for `Or`s containing read only fetches.

File: crates/bevy_ecs/src/system/mod.rs
Comment (cart) [reply crates/bevy_ecs/src/system/mod.rs]: Removed!

Comment (cart) [reply crates/bevy_ecs/src/system/mod.rs]: (and added a new trybuild equivalent of the `mut_overlap_test`)

**Pull request closed without merging**
